src/eth_provider/provider.rs

```diff
@@ -721,12 +720,40 @@ where
         block_id: impl Into<Option<BlockId>>,
     ) -> EthProviderResult<starknet::core::types::BlockId> {
         match block_id.into() {
-            Some(BlockId::Hash(hash)) => {
-                Ok(starknet::core::types::BlockId::Number(self.hash_into_block_number(hash).await?.to::<u64>()))
-            }
-            Some(id) => {
-                let eth_block_id = EthBlockId::new(id);
-                Ok(eth_block_id.try_into()?)
+            Some(BlockId::Hash(hash)) => Ok(EthBlockId::new(BlockId::Hash(hash)).try_into()?),
+            Some(BlockId::Number(number_or_tag)) => {
+                // There is a need to separate the BlockNumberOrTag case into three subcases
+                // because pending Starknet blocks don't have a number.
+                // 1. The block number corresponds to a Starknet pending block, then we return the pending tag
+                // 2. The block number corresponds to a Starknet sealed block, then we return the block number
+                // 3. The block number is not found, then we return an error
+                if let BlockNumberOrTag::Number(number) = number_or_tag {
+                    let filter = doc! {};
+                    let sort = doc! { "header.number": -1 };
+                    let header: Option<StoredHeader> = self.database.get_one("headers", filter, sort).await?;
```

Comment:
Is this really correct? So you get the latest block. If the latest block number == the number you query and the latest block hash == "0x0" => return Pending. Else return number. But if number is huge, above the latest block number, you will still return this number? 
I think we need to add a test to verify the logic and you should probably do something like
```rust
let header = self.header(BlockHashOrNumber::Number(number))?.ok_or(EthProviderError::ValueNotFound("Block".to_string()));
if header.hash == FixedBytes::ZERO {
Ok(starknet::core::types::BlockId::Tag(starknet::core::types::BlockTag::Pending))
} else {
Ok(starknet::core::types::BlockId::Number(number))
}
```
This is a bit pseudo code so probs won't compile.